salt/requisite.py

```diff
@@ -0,0 +1,412 @@
+LowChunk = dict[str, Any]
+
+
+def _gen_tag(low: LowChunk) -> str:
+    """
+    Generate the a unique identifer tag string from the low data structure
+    """
+    return "{0[state]}_|-{0[__id__]}_|-{0[name]}_|-{0[fun]}".format(low)
+
+
+def trim_req(req: dict[str, Any]) -> dict[str, Any]:
+    """
+    Trim any function off of a requisite reference
+    """
+    reqfirst, valfirst = next(iter(req.items()))
+    if "." in reqfirst:
+        return {reqfirst.split(".", maxsplit=1)[0]: valfirst}
+    return req
+
+
+class RequisiteType(str, Enum):
+    """Types of direct requisites"""
+
+    @staticmethod
+    def _generate_next_value_(name: str,
+                              start: int,
+                              count: int,
+                              last_values: list[Any]) -> tuple[str, int]:
+        return name.lower(), count
+
+    def __new__(cls, value, weight):
+        member = str.__new__(cls, value)
+        member._value_ = value
+        member.weight = weight
+        return member
+
+    def __init__(self, value, weight):
+        super().__init__()
+        self._value_ = value
+        self.weight = weight
+
+    # The items here are listed in order of precedence for determining
+    # the order of execution, so do not change the order unless you
+    # are intentionally changing the precedence
+    ONFAIL = auto()
+    ONFAIL_ANY = auto()
+    ONFAIL_ALL = auto()
+    REQUIRE = auto()
+    REQUIRE_ANY = auto()
+    ONCHANGES = auto()
+    ONCHANGES_ANY = auto()
+    WATCH = auto()
+    WATCH_ANY = auto()
+    PREREQ = auto()
+    PREREQUIRED = auto()
+    LISTEN = auto()
+
+
+class DependencyGraph:
+    """
+    Class used to track dependencies (requisites) among salt states.
+
+    This class utilizes a Directed Acyclic Graph to determine the
+    ordering of states. The nodes represent the individual states that
+    can be depended on and edges represent the types of requisites
+    between the states.
+    """
+
+    def __init__(self) -> None:
+        self.dag = nx.MultiDiGraph()
+        # a mapping to node_id to be able to find nodes with
+        # specific state type (module name), names, and/or IDs
+        self.nodes_lookup_map: dict[tuple[str, str], set[str]] = {}
+        self.sls_to_nodes: dict[str, set[str]] = {}
+
+    def _add_prereq(self,
+                    node_tag: str,
+                    req_type: RequisiteType,
+                    req_tag: str):
+        prereq_chunk = self.dag.nodes[req_tag]["chunk"]
+        prereq_chunk["__prereq__"] = True
+        prereq_check_node = f"prereq_check_{req_tag}"
+        self.dag.add_node(prereq_check_node,
+                          chunk=prereq_chunk,
+                          state=prereq_chunk["state"])
+        self.dag.add_edge(prereq_check_node, node_tag,
+                          RequisiteType.PREREQUIRED)
+        for u, _, key, data in self.dag.in_edges(
```

Comment:
```suggestion
        for u, _, key, data in self.dag.in_edges(
```
What is `u` here?
Can we get a name that represents what this is?

Reply:
1. This is common naming for usage of the graph library for the first node in the vertex `(u, v)`. I will trying to figure out a more domain specific name.

Reply:
> I will trying to figure out a more domain specific name.

Thanks